File: 0148-sort-list/0148-sort-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
     public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode head = new ListNode(0) ; 
        ListNode temp = head ; 
        while(list1!=null&&list2!=null){
            if(list1.val<list2.val){
                temp.next = new ListNode(list1.val) ;
                temp = temp.next ;
                list1 = list1.next;
            }
            else{
                temp.next = new ListNode(list2.val) ; 
                temp = temp.next ;
                list2 = list2.next ;
            }
        }
        while(list1!=null){
            temp.next = new ListNode(list1.val) ;
            temp = temp.next ;
            list1 = list1.next ;
        }
        while(list2!=null){
            temp.next = new ListNode(list2.val) ;
            temp = temp.next ;
            list2 = list2.next ;
        }
        head = head.next ; 
        return head ;
    }
    
    

     
  
    public ListNode sortList(ListNode head) {
            if(head==null||head.next==null)
            return head ;
             ListNode midPrev = null;
        ListNode slow = head;
        ListNode fast = head;
        while (fast != null && fast.next != null) {
            midPrev = slow;
            slow = slow.next;
            fast = fast.next.next;
        }
         ListNode lis =   sortList(midPrev.next) ;
          midPrev.next = null;    
          ListNode list =  sortList(head) ;
          return mergeTwoLists(lis , list) ;
    }
}

File: 0148-sort-list/0148-sort-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
#define Node struct ListNode* 

Node newNode(int n){
    Node  node = malloc(sizeof(Node)) ; 
    node->val = n ;
    node->next = NULL ;
    return node;
}

Node merge(Node temp1, Node temp2){
    if(temp1==NULL){
        return temp2;
    }
    if(temp2==NULL){
        return temp1 ;
    }
    if(temp1->val<temp2->val){
        temp1->next = merge(temp1->next , temp2);
        return temp1 ;
    }
    else{
        temp2->next = merge(temp1,temp2->next);
        return temp2;
    }
}
Node mid(Node temp){
    Node slow = temp ;
    Node  prev = NULL ;
    Node fast = temp ; 
    while(fast!=NULL&&fast->next!=NULL){
        fast = fast->next->next ; 
        prev  = slow ;
        slow = slow->next ;
    }
    if(prev!=NULL)
    prev->next = NULL ;
    return slow ;
}
Node  sort(Node temp){
    if(temp==NULL||temp->next==NULL){
        return temp;
    }
    return merge(sort(temp),sort(mid(temp))) ;
}
struct ListNode* sortList(struct ListNode* head) {
    return sort(head);
}